### PR TITLE
Wicket 4201 improved page provider

### PR DESCRIPTION
**Pull request by martin-g**
Creating a Pull Request so we can comment on the changes.

### CODE REVIEW

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java [hunk 1]
```diff
@@ -54,15 +52,14 @@
 
 	private final Integer pageId;
 
-	private IPageSource pageSource;
-
-	private transient IRequestablePage pageInstance;
-	private boolean pageInstanceIsFresh;
+	private transient IPageSource pageSource;
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: IMO using `transient` is just delaying the problem.
`PageProvider` should not be serializable (IClusterable).
If it is serialized and then deserialized then most probably some call with fail with NullPointerException because of those transient fields.

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java [hunk 2]
```diff
@@ -197,26 +194,49 @@ else if (isNewPageInstance() == false)
...
+	 * 
+	 * @return {@code true} iff the page instance held by this provider was instantiated by the
+	 *         provider
+	 */
+	@Override
+	public final boolean doesProvideNewPage()
+	{
+		if (provision == null)
+		{
+			throw new IllegalStateException("Page instance not yet resolved");
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: This doesn't look nice.
At https://issues.apache.org/jira/browse/WICKET-4201?focusedCommentId=15853108&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15853108 you say 
 ` removed the PageExpiredException from PageProvider since it breaks the class coherence. It's designed to provide pages and related metadata (such as if the page was expired included), not to change the application flow on application exceptions.` but this runtime exception does exactly this.
IMO it would be better to return a three state result: `NEW`, `OLD` and `UNKNOWN`

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java [hunk 3]
```diff
@@ -349,6 +306,11 @@ public void detach()
 	 */
 	public void setPageSource(IPageSource pageSource)
 	{
+		if (provision != null)
+		{
+			throw new IllegalStateException(
+				"A provision was already been done. The provider can be forcefully detached or a new one needs to be used to provide using this page source.");
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: `"A provision has already been done. The provider can be forcefully detached or a new one has to be created be be able to use this page source."` ?

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java [hunk 4]
```diff
@@ -388,49 +350,95 @@ public Integer getRenderCount()
...
+
+				failedToFindStoredPage = page == null;
+			}
+
+			return this;
+		}
+
+		void detach()
+		{
+			if (page != null)
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: IMO this should just reset its own member fields.
Why do you think it should detach the page too ?

File: wicket-core/src/test/java/org/apache/wicket/core/request/mapper/MountedMapperTest.java [hunk 5]
```diff
@@ -714,7 +714,7 @@ public boolean isNewPageInstance()
 	@Test
 	public void placeholderEncode4()
 	{
-		PageProvider provider = new PageProvider(new MockPage())
+		PageProvider provider = new PageProvider(MockPage.class)
```
Comment (martin-g) [inline wicket-core/src/test/java/org/apache/wicket/core/request/mapper/MountedMapperTest.java]: Why this change is needed ?
This is not the same as before.

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java (hunk 2 above)
Comment (pedrosans) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: Looking bad indeed, I fixed later during my work on the branch: https://github.com/apache/wicket/pull/210/commits/b64f2b4b2cb50a594cabb50a0c1eeca43485f6b9#diff-36b11db0d92a540368a080c520de953aL177

The three states looks overcomplicated. There's no place in Wicket core needing to test such flag for the value "OLD".

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java (hunk 3 above)
Comment (pedrosans) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: My bad, I meant to say: "to be able to **change** the page source". I will improve this exception message.

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java (hunk 4 above)
Comment (pedrosans) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: Good point, I just kept the old behaviour here.

File: wicket-core/src/test/java/org/apache/wicket/core/request/mapper/MountedMapperTest.java (hunk 5 above)
Comment (pedrosans) [inline wicket-core/src/test/java/org/apache/wicket/core/request/mapper/MountedMapperTest.java]: Not the same and the change is needed. This test follows up by changing the page provider which will have no impact on the provider that already holds an instance.

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java (hunk 2 above)
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: For me it is strange that a method with return type `boolean` returns either `true` or throws an exception.

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java (hunk 4 above)
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: Well, if the old code detached it then let's keep it.
There is an open ticket in JIRA that sometimes a Page is detached twice in the same request cycle, that's why I've asked.

File: wicket-core/src/test/java/org/apache/wicket/core/request/mapper/MountedMapperTest.java (hunk 5 above)
Comment (martin-g) [inline wicket-core/src/test/java/org/apache/wicket/core/request/mapper/MountedMapperTest.java]: I didn't understand you.
```java
/**
	 * WICKET-5247 page instantiated without required parameters won't be mapped
	 */
	@Test
	public void placeholderEncode4()
	{
		PageProvider provider = new PageProvider(new MockPage())
		{
			@Override
			public boolean isNewPageInstance()
			{
				return false;
			}
		};
		provider.setPageSource(context);
		IRequestHandler handler = new RenderPageRequestHandler(provider);
		Url url = placeholderEncoder.mapHandler(handler);
		assertNull(url);
	}
```
The test setups a PageProvider and uses it. There is no replacement after.
By making this change I think we don't test the same as what broke in WICKET-5247.

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java (hunk 1 above)
Comment (pedrosans) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: It can lead to inconsistent behaviour indeed, but not NPE since it will default to the application one. I will support the removal of IClusterable interface. From what I looked in the code, page provider is just a core's internal communicator of pages construction and fetch logic. It's used during the work to handle a page request and never shared on the session. I would bet that at some point a developer needed ou wanted a page provider as a page attribute and ended up marking it as serializable.

File: wicket-core/src/test/java/org/apache/wicket/core/request/mapper/MountedMapperTest.java (hunk 5 above)
Comment (pedrosans) [inline wicket-core/src/test/java/org/apache/wicket/core/request/mapper/MountedMapperTest.java]: My bad again, I meant to say: "... test follows up by changing the page **source** ..."
The line: 
      provider.setPageSource(context);
was setting up a page source that was never used because the page provider already had a page instance (the one passed as a parameter in the page provider). 

I don't know why the test sets the page source, but either way a page with no parameters will be provided making it to test the same thing. Maybe we can just remove the custom page source.

File: wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java (hunk 2 above)
Comment (pedrosans) [inline wicket-core/src/main/java/org/apache/wicket/core/request/handler/PageProvider.java]: oh, just just saw that this is a comment on the doesProvideNewPage method (github interface didn't show the method signature and I assumed we were talking about PageProvider#getPageInstance, where I indeed removed the illegal state exception. PageProvider#doesProvideNewPage can return true or false, but regardless I agreed that it shouldn't throw a runtime exception. Thanks for pointing it, will fix next.

**Comment by martin-g** [comment]
@pedrosans I think this could be merged to `master`!
